src/CodeGen_Hexagon.cpp

```diff
@@ -2696,32 +2696,37 @@ void CodeGen_Hexagon::visit(const Load *op) {
     const Ramp *ramp = op->index.as<Ramp>();
     const IntImm *stride = ramp ? ramp->stride.as<IntImm>() : NULL;
     if (ramp && stride && stride->value == 1) {
-      int width = ramp->lanes;
-      ModulusRemainder mod_rem = get_alignment_info(ramp->base);
+      int lanes = ramp->lanes;
+      // int alignment = op->type.bytes(); // The size of a single element
+      int native_vector_bytes = native_vector_bits() / 8;
 
-      int alignment_required = CPICK(128, 64);
-      if (width != alignment_required) {
-        // This will happen only under two cases.
-        // 1. We are loading a partial vector.
-        // 2. We are loading a type other than u8 or i8.
-        // FIXME: PDB: Relax the second condition.
+      // We are loading a partial vector. if we are, default to vanilla codegen.
+      if (lanes * op->type.bytes() != native_vector_bytes) {
         CodeGen_Posix::visit(op);
         return;
       }
+      // At this point we are satisfied that we are loading a native vector.
+      ModulusRemainder mod_rem = get_alignment_info(ramp->base);
       if (mod_rem.modulus == 1 &&
           mod_rem.remainder == 0) {
         // We know nothing about alignment. Just fall back upon vanilla codegen.
         CodeGen_Posix::visit(op);
         return;
       }
-      if (!possibly_misaligned && mod_rem.modulus == alignment_required) {
+      // ModulusRemainder tells us if something can be written in the
+      // form
+      // (ModulusRemainder.modulus * c1) + ModulusRemainder.remainder.
+      // So for us to be able to generate an aligned load, ramp->base
+      // should be
+      // (lanes * c1) + c2.
+      if (!possibly_misaligned && !(mod_rem.modulus % lanes)) {
         if (mod_rem.remainder == 0) {
           // This is a perfectly aligned address. Vanilla codegen can deal with
           // this.
           CodeGen_Posix::visit(op);
           return;
         } else {
-          Expr base = ramp->base;
+          Expr base = simplify(ramp->base);
```

Comment:
Yes, I believe the source of that was this  in CodeGen_LLVM.cpp when we are lowering a load with stride == 2.
`else if (ramp && stride && stride->value == 2) {
            // Load two vectors worth and then shuffle
            Expr base_a = ramp->base, base_b = ramp->base + ramp->lanes;
`

base_b needs to be wrapped inside a simplify.